src/audio/dp_queue.c

```diff
@@ -0,0 +1,391 @@
+
+	bool hw_params_configured;
+	struct sof_audio_stream_params audio_stream_params;
+
+	struct k_spinlock lock;
+};
+
+union lock_key {
+	k_spinlock_key_t spinlock;
+	int irq_lock;
+};
+
+static inline bool dp_queue_is_shared(struct dp_queue *dp_queue)
+{
+	return !!(dp_queue->flags & DP_QUEUE_MODE_SHARED);
+}
+
+static inline uint8_t __sparse_cache *dp_queue_buffer_end(struct dp_queue *dp_queue)
+{
+	return dp_queue->data_buffer + dp_queue->data_buffer_size;
+}
+
+static inline struct dp_queue *dp_queue_get_from_sink(struct sof_sink *sink)
+{
+	return container_of(sink, struct dp_queue, sink_api);
+}
+
+static inline struct dp_queue *dp_queue_get_from_source(struct sof_source *source)
+{
+	return container_of(source, struct dp_queue, source_api);
+}
+
+static inline void dp_queue_invalidate_shared(struct dp_queue *dp_queue,
+					      void __sparse_cache *ptr, uint32_t size)
+{
+	/* wrap-around? */
+	if ((uintptr_t)ptr + size > (uintptr_t)dp_queue_buffer_end(dp_queue)) {
+		/* writeback till the end of circular buffer */
+		dcache_invalidate_region
+			(ptr, (uintptr_t)dp_queue_buffer_end(dp_queue) - (uintptr_t)ptr);
+		size -= (uintptr_t)dp_queue_buffer_end(dp_queue) - (uintptr_t)ptr;
+		ptr = (__sparse_force void __sparse_cache *)dp_queue->data_buffer;
+	}
+	/* invalidate rest of data */
+	dcache_invalidate_region(ptr, size);
+}
+
+static inline union lock_key dp_queue_lock(struct dp_queue *dp_queue)
+{
+	/* use cross-core spinlock in case of shared queue
+	 * When shared, dp_queue structure is located in not cached memory
+	 * as required by spinlock
+	 */
+	union lock_key key;
+
+	if (dp_queue_is_shared(dp_queue))
+		key.spinlock = k_spin_lock(&dp_queue->lock);
+	else
+		/* use faster irq_lock in case of not shared queue (located in cached mem) */
+		key.irq_lock = irq_lock();
+
+	return key;
+}
+
+static inline void dp_queue_unlock(struct dp_queue *dp_queue, union lock_key key)
+{
+	if (dp_queue_is_shared(dp_queue))
+		k_spin_unlock(&dp_queue->lock, key.spinlock);
+	else
+		irq_unlock(key.irq_lock);
+}
+
+struct sof_sink *dp_queue_get_sink(struct dp_queue *dp_queue)
+{
+	return &dp_queue->sink_api;
+}
+
+struct sof_source *dp_queue_get_source(struct dp_queue *dp_queue)
+{
+	return &dp_queue->source_api;
+}
+
+static size_t dp_queue_get_free_size(struct sof_sink *sink)
+{
+	struct dp_queue *dp_queue = dp_queue_get_from_sink(sink);
+
+	return dp_queue->free_space;
+}
+
+static int dp_queue_get_buffer(struct sof_sink *sink, size_t req_size,
+			       void **data_ptr, void **buffer_start, size_t *buffer_size)
+{
+	struct dp_queue *dp_queue = dp_queue_get_from_sink(sink);
+
+	/* dp_queue_get_free_size will return free size with adjustment for cacheline if needed */
+	if (req_size > dp_queue_get_free_size(sink))
+		return -ENODATA;
+
+	/* no need to lock, just reading data that may be modified by commit_buffer only
+	 *
+	 * note! a sparse warning will be generated here till
+	 * https://github.com/thesofproject/sof/issues/8006 is implemented
```

Comment:
can't help it now, it is a soft dependency and I hope we can implement it soon enough

Reply:
it can be avoided by adding type-casts

Reply:
yes, it can. Yet the PR implementing 8006 is on its way, probably early next week. Either a comment (to be removed) or cast (to be removed) 
Probably none of them will be merged